src/memory-usage.ts

```diff
@@ -0,0 +1,109 @@
+// Implements the "memory usage" feature.
+// When the server advertises `memoryUsageProvider`, a command
+// (clangd.memoryUsage) is available (context variable:
+// clangd.memoryUsage.supported). It sends the $/memoryUsage request and
+// displays the result in a tree view (clangd.memoryUsage) which becomes visible
+// (context: clangd.memoryUsage.hasData)
+
+import * as vscode from 'vscode';
+import { window } from 'vscode';
+import * as vscodelc from 'vscode-languageclient/node';
+
+import {ClangdContext} from './clangd-context';
+
+export function activate(context: ClangdContext) {
+  const feature = new ASTFeature(context);
+  context.client.registerFeature(feature);
+}
+
+// LSP wire format for this clangd feature.
+interface NoParams {}
+interface WireTree {
+  _self: number;
+  _total: number;
+  [child: string]: WireTree|number|string;
+}
+export const RequestType =
+    new vscodelc.RequestType<NoParams, WireTree, void, void>('$/memoryUsage');
+
+// Internal representation that's a bit easier to work with.
+interface InternalTree {
+  title: string;
+  total: number;
+  self: number;
+  isFile: boolean; // different icon, default collapsed
+  children: InternalTree[];
+}
+function convert(m: WireTree, title: string): InternalTree {
+  const slash = Math.max(title.lastIndexOf('/'), title.lastIndexOf('\\'));
+  return {
+    title: title.substr(slash + 1), // display basename only
+    isFile: slash >= 0,
+    total: m._total,
+    self: m._self,
+    children: Object.keys(m)
+                  .filter(x => !x.startsWith('_'))
+                  .map(e => convert(m[e] as WireTree, e))
+                  .sort((x, y) => x.total - y.total)
+                  .reverse(),
```

Comment:
why not sort in the other way directly? (or why not sort based on title)

Reply:
You're right. The intention was to sort by [total, title] but I messed it up.

The code to sort by a compound key is awkward in JS, but sort() is stable, so we sort the titles first and then the totals.